mlir/lib/Dialect/MQTOpt/Transforms/Transpilation/sc/RoutingPass.cpp

```diff
@@ -0,0 +1,1071 @@
+  if (hasElse) {
+    rewriter.mergeBlocks(&op.getElseRegion().front(),
+                         &ifOp.getElseRegion().front());
+  }
+
+  if (nresults > 0) {
+    rewriter.replaceOp(op, ifOp.getResults().take_front(nresults));
+  } else {
+    rewriter.eraseOp(op);
+  }
+
+  /// Prepare stack.
+  stack.duplicateTopState(); // Else
+  stack.duplicateTopState(); // Then
+
+  /// Forward results for all hardware qubits.
+  LayoutState& stateBeforeIf = stack.getStateAtDepth(IF_PARENT_DEPTH);
+  for (std::size_t i = 0; i < qubits.size(); ++i) {
+    const Value in = stateBeforeIf.getHardwareQubits()[i];
+    const Value out = ifOp->getResult(i);
+    stateBeforeIf.remapQubitValue(in, out);
+  }
+
+  return WalkResult::advance();
+}
+
+/**
+ * @brief Defines the end of a region (and hence routing state) defined by a
+ * branching op. Consequently, we pop the region's state from the stack.
+ *
+ * Restores layout by uncomputation and replaces (invalid) yield.
+ *
+ * The results of the yield op are extended by the missing hardware
+ * qubits, similarly to the 'for' and 'if' op. This is only possible
+ * because we restore the layout - the mapping from hardware to program
+ * qubits (and vice versa).
+ *
+ * Using uncompute has the advantages of (1) being intuitive and
+ * (2) preserving the optimally of the original SWAP sequence.
+ * Essentially the better the routing algorithm the better the
+ * uncompute. Moreover, this has the nice property that routing
+ * a 'for' of 'if' region always requires 2 * #(SWAPs required for region)
+ * additional SWAPS.
+ */
+WalkResult handleYield(scf::YieldOp op, StateStack& stack,
+                       PatternRewriter& rewriter) {
+  if (!isa<scf::ForOp>(op->getParentOp()) &&
+      !isa<scf::IfOp>(op->getParentOp())) {
+    return WalkResult::skip();
+  }
+
+  for (const auto [programIdx0, programIdx1] :
+       llvm::reverse(stack.getHistory())) {
+    const Value qIn0 = stack.topState().lookupProgram(programIdx0);
+    const Value qIn1 = stack.topState().lookupProgram(programIdx1);
+
+    auto swap = createSwap(op->getLoc(), qIn0, qIn1, rewriter);
+    const auto [qOut0, qOut1] = getOuts(swap);
+
+    rewriter.setInsertionPointAfter(swap);
+    replaceAllUsesInRegionAndChildrenExcept(
+        qIn0, qOut1, swap->getParentRegion(), swap, rewriter);
+    replaceAllUsesInRegionAndChildrenExcept(
+        qIn1, qOut0, swap->getParentRegion(), swap, rewriter);
+
+    stack.topState().swapProgramIndices(qIn0, qIn1);
+    stack.topState().remapQubitValue(qIn0, qOut0);
+    stack.topState().remapQubitValue(qIn1, qOut1);
+  }
+
+  assert(llvm::equal(stack.topState().getCurrentLayout(),
+                     stack.getStateAtDepth(1).getCurrentLayout()) &&
+         "layouts must match after restoration");
+
+  const SmallVector<Value> newResults(stack.topState().getHardwareQubits());
+  rewriter.replaceOpWithNewOp<scf::YieldOp>(op, newResults);
+
+  stack.pop();
+
+  return WalkResult::advance();
+}
```

Comment:
Again, nothing for this PR, but maybe for a slightly sooner follow-up:
This is not necessarily optimal. In many cases, there may be better ways to revert from the final permutation back to the initial permutation than to simply play back the SWAPs in reverse.
At the top of my head, I can't quite remember how hard the classical problem of permutation synthesis out of locally restricted SWAPs is, but I would guess that a different heuristic would perform a bit better here.
Such a heuristic already exists in MQT Core as part of the DD package: https://github.com/munich-quantum-toolkit/core/blob/5058297f58fb7a2efdea9c2ca6ef6caf39edfed1/include/mqt-core/dd/Operations.hpp#L248-L304
This transforms from one permutation to another in a greedy fashion without connectivity constraints.

Reply:
Thinking about it. This could actually be part of the router implementation.

Reply:
Yeah, there could be different strategies for this that one can choose from.

Reply:
Absolutely! Either way it would also be pretty nice to get rid of the SWAP history for the naive approach.